uaclient/contract.py

```diff
@@ -165,12 +165,19 @@ def detach_machine_from_contract(
         self, machine_token: str, contract_id: str, machine_id: str = None
     ) -> "Dict":
         """Report the attached machine should be detached from the contract."""
-        return self._request_machine_token_update(
-            machine_token=machine_token,
-            contract_id=contract_id,
-            machine_id=machine_id,
-            detach=True,
-        )
+        try:
+            return self._request_machine_token_update(
+                machine_token=machine_token,
+                contract_id=contract_id,
+                machine_id=machine_id,
+                detach=True,
+            )
+        except ContractAPIError as e:
+            if detach and e.code == 403 and "token invalid for machine" in e:
```

Comment:
It seems that the `detach` variable does not exist.

After we fix that, what would happen to PRO snapshot ? It will not be able to perfom any `uaclient` operation until it cleans the `auclient` artifacts, right ?

Reply:
Right sorry I moved where this code from where I originally wrote it in _request_machine_token_update and decided it was better handled here in detach_machine_from_contract. So, yes just dropping that detach check would work.